chain/epoch_manager/src/validator_selection.rs

```diff
@@ -97,39 +110,67 @@ pub fn proposals_to_epoch_info(
-                hash_map::Entry::Vacant(entry) => {
-                    let validator_id = i as ValidatorId;
-                    entry.insert(validator_id);
-                    shard_validator_ids.push(validator_id);
-                    all_validators.push(validator);
-                    i += 1;
-                }
-                // Validators which have an entry in the validator_to_index map
-                // have already been inserted into `all_validators`.
-                hash_map::Entry::Occupied(entry) => {
-                    let validator_id = *entry.get();
-                    shard_validator_ids.push(validator_id);
+    let chunk_producers_settlement = checked_feature!(
+        "protocol_feature_chunk_only_producers",
+        ChunkOnlyProducers,
+        next_version,
+        {
+            let minimum_validators_per_shard =
+                epoch_config.validator_selection_config.minimum_validators_per_shard as usize;
+            let shard_assignment =
+                assign_shards(chunk_producers, num_shards as usize, minimum_validators_per_shard)
+                    .map_err(|_| EpochError::NotEnoughValidators {
+                    num_validators: num_chunk_producers as u64,
+                    num_shards,
+                })?;
+
+            let mut chunk_producers_settlement: Vec<Vec<ValidatorId>> =
+                shard_assignment.iter().map(|vs| Vec::with_capacity(vs.len())).collect();
+            let mut i = all_validators.len();
+            for (shard_validators, shard_validator_ids) in
+                shard_assignment.into_iter().zip(chunk_producers_settlement.iter_mut())
+            {
+                for validator in shard_validators {
+                    debug_assert_eq!(i, all_validators.len());
+                    match validator_to_index.entry(validator.account_id().clone()) {
+                        hash_map::Entry::Vacant(entry) => {
+                            let validator_id = i as ValidatorId;
+                            entry.insert(validator_id);
+                            shard_validator_ids.push(validator_id);
+                            all_validators.push(validator);
+                            i += 1;
+                        }
+                        // Validators which have an entry in the validator_to_index map
+                        // have already been inserted into `all_validators`.
+                        hash_map::Entry::Occupied(entry) => {
+                            let validator_id = *entry.get();
+                            shard_validator_ids.push(validator_id);
+                        }
+                    }
                 }
             }
+            chunk_producers_settlement
+        },
+        {
+            if chunk_producers.is_empty() {
+                // All validators tried to unstake?
+                return Err(EpochError::NotEnoughValidators { num_validators: 0u64, num_shards });
+            }
+            let mut id = 0usize;
```

Comment:
I'd add a comment that old algorithm was doing a simple round-robin over the block_producers